### ISSUE TITLE
VIP: override storage slots using @ slot annotations

### ISSUE THREAD

**Issue by banteg** [vyperlang/vyper#3556]
## Simple Summary
- deprecate `--storage-layout-file` option in favor of storage slot override syntax
- introduce `name: type @ slot` syntax for storage slot override

## Motivation
- overriding all storage slots is cumbersome and rarely needed. partially overriding storage slots can make it possible to support certain eips like [eip-1967](https://eips.ethereum.org/EIPS/eip-1967) that require certain positions in a very clear manner.
- `--storage-layout-file` is little known and has poor tool support. neither of vyper's own standard json input (until recently), etherscan verification, ape-vyper or vvm have support for it. having it in the contract itself makes the support from upstream tools follow naturally.

## Specification
- deprecate `--storage-layout-file` option of the compiler
- introduce new syntax to partially override storage layout using the `@` aka `matmul` operator, which doesn't require any special parsing since it's a native python operator. the slot value must be either of `int` or `hex int` literal.
- example usage
```python
implementation: public(address) @ 0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc
admin: public(address) @ 0xb53127684a568b3173ae13b9f8a6016e243e63b6e8ee1178d6a717850b5d6103
```
- alternatives considered. note that this example needs special handling because `@` has lower precedence than `-`.
```python
implementation: public(address) @ keccak256("eip1967.proxy.implementation") - 1
admin: public(address) @ keccak256("eip1967.proxy.admin") - 1
```

## Backwards Compatibility
this vip deprecates functionality in favor of new design. it doesn't introduce backwards incompatibility for contracts that don't use storage slot overrides.

## Dependencies
this vip doesn't have any dependencies blocking it.

## References
this vip supersedes https://github.com/vyperlang/vyper/issues/2572

## Copyright
Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/)

**Comment by charles-cooper** [comment]
my biggest concern here is "partially overriding storage slots". it seems like a footgun, where some slots are overridden (ostensibly for the purpose of making a contract upgradeable) and some other slots are not, and then they get trampled during a contract upgrade.

**Comment by 0x8a8a** [comment]
Reviving this discussion.

Maybe the operand to the `@` operator, can be a hex, number, or string. Hex and number are self explanatory. But for a string, can use the same erc7201 method to calculate the desired slot. formula being: `erc7201(id: string) = keccak256(keccak256(id) - 1) & ~0xff`

#### Alternate design idea

As an alternative to using the `@` operator. What if it used the natspec convention descrbed in eip-7201 ??

Example using [examples/storage/storage.vy](https://github.com/vyperlang/vyper/blob/68b68c4b30c5ef2f312b4674676170b8a6eaa316/examples/storage/storage.vy)

```
# pragma version >0.4.1

# @custom:storage-location erc7201:example.x
x: public(uint256)

"""
@custom:storage-location erc7201:example.y
"""
y: public(uint256)

@deploy
def __init__(x: uint256, y: uint256):
  self.x = x
  self.y = y

@external
def set(x: uint256, y: uint256):
  self.x = x
  self.y = y
```

Major downside to this is that it's using comments to influence the compiler which I can't imagine would be a good design principle. Having two different compiler outputs because of comments seems bad.

### References

- https://eips.ethereum.org/EIPS/eip-7201

**Comment by charles-cooper** [comment]
> Having two different compiler outputs because of comments seems bad.

yes, i agree. the exception here is pragmas, which set compiler mode, but those have specific syntax, are handled very early in the parse stage and affect the whole file.

**Comment by Sporarum** [comment]
I don't think we should add some magic syntax with an operator on the type
Especially since storage slots are properties of the variable, and not of the value (which types tend to be)

Decorators seem appropriate in this case, since they annotate properties of the variable:
```
@storage_slot(0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc)
implementation: public(address)

@storage_slot(0xb53127684a568b3173ae13b9f8a6016e243e63b6e8ee1178d6a717850b5d6103)
admin: public(address)
```

It also has the upside that it is a bit more self-documenting than a symbolic operators, which are hard to search online

For 0.5.0 I think we should only support hex-literals, to provide a migration path for `--storage-layout-file`, we can add better UX like `keccak256("eip1967.proxy.implementation") - 1` in 0.5.1
And also only forbid them from interfaces for now, and we'll make a clean system for that also in 0.5.0

**Comment by Sporarum** [comment]
I think the big question is: Who should own the storage slots ?
(in the sense, who has final say)

## Modules

The current proposal seems to favor modules having final say over their slots. This creates issues when two modules share the same slot, as it might be on purpose (memory sharing, [EIP1967](https://eips.ethereum.org/EIPS/eip-1967)) or accidental (modules both used the same slot, but different semantic information is supposed to reside there).
We could forbid this at compile-time, but it would mean modules need to avoid each other, which is not ideal given modules should ideally behave independently (i.e. with no knowledge of each other).

## The Top-level Module

The other option is to have the top-level module have final say. This also creates issues as there is no syntactical markers for top-level modules, it also means the situation is not much different from the current `--storage-layout-file`. The only differences being: 1) it resides in source code, and 2) it can be partial.

1) is both an upside and a downside, as it means it's harder to delegate storage slots to tooling (imagine a git-like tool for upgradeable contracts, which handles computing the diff and generating a storage layout file automatically from it)
2) can also be done by making the storage layout file more flexible. We can make missing entries in the file resolve according to the default vyper rules.

An additional issue is that it somewhat hurts the modularity of modules, given the top-level module can no longer fully delegate logic to it's children modules.

**Comment by charles-cooper** [comment]
Hmm that's a good way of framing it, I think it is similar to the problem of overrides (who has final say on which is the concrete implementation) and I think we can maybe resolve using the same machinery

**Comment by Sporarum** [comment]
I think the overrides situation was simpler, because naming was local (at the interface between two modules), whereas storage slots are global.
So it's hard to find a solution which keeps the module logic local, while being globally aware.
I think this is one of the strengths of https://github.com/vyperlang/vyper/issues/1733 (or other solutions of the kind "module is a eip-7201 namespace")

**Comment by Sporarum** [comment]
The simplest solution I can come up with is the following:

```python
import module_a
import module_b

__storage_slots__ = {
  module_a.foo: <storage slot here>
  module_b.module_c.bar: <storage slot here>
}
```
With the following restrictions:
1. Only allowed on top-level module (But how do we allow `other_contract.__interface__` for one which uses storage slots ?)
2. Can be partial, vyper uses default strategy for remaining slots
3. Compile-time error if slots overlap (can be foolproof as long as only bounded types are allowed in storage)
4. The storage slot must inline to an int literal, potentially we can force inline `pure` functions to make it work, allows:
```python
implementation: address

@pure
def keccak256(s: String[INF]) -> uint256:
    # logic here

__storage_slots__ = {
  implementation: keccak256("eip1967.proxy.implementation") - 1
}
```


It's basically just the existing system but instead of a side-car file it resides in the source code of the top-level module, and is allowed to be partial.
By allowing `__storage_slots__` to be located anywhere in the top-level module, this helps with the inherent tooling downsides, by allowing tooling to just append it at the very end of the file


(Note that I find my current proposal syntactically much less elegant than the original, but that elegance was hiding some underlying flaws, which are hopefully absent from this one)

**Comment by Sporarum** [comment]
> 3\. Compile-time error if slots overlap (can be foolproof as long as only bounded types are allowed in storage)

Actually not sure this is true in the presence of hashmaps

**Comment by Sporarum** [comment]
@charles-cooper proposed using nesting:
```python
__storage_slots__:
    module_a:
        foo: <slot>
    module_b:
        bar: <slot>
```

This probably scales better when we want to define a lot of slots:
```python
__storage_slots__ = {
    module_a.foo: <slot>,
    module_a.bar: <slot>,
    module_a.baz: <slot>,
}
# vs
__storage_slots__:
    module_a:
        foo: <slot>
        bar: <slot>
        baz: <slot>
```

If we do this, it will no longer be a magic value, so we can drop the dunder:
```python
storage_slots:
    module_a:
        foo: <slot>
    module_b:
        bar: <slot>
```
